File: CPP_DSApractice-main/Nqueens.cpp


#include <iostream>
using namespace std;
void printboard(int a[][20],int n){
    for (int i = 0;i<n;i++){
        for(int j=0;j<n;j++){
            if(a[i][j]==1){
                cout<<"Q"<<" ";
            }
            else{
                cout<<"0"<<" ";
            }
        }
        cout<<endl;
    }
}
bool canplace(int a[][20],int n,int i,int j){
    //check up
    for(int x=i;x>=0;x--){
        if(a[x][j]==1){
            return false;
        }
    }
    
    //check diagonal left
    for(int x=i,y=j;x>=0 && y>=0;x--,y--){
        if(a[x][y]==1){
            return false;
        }
    }
    
    //check diagonal right
    for(int x=i,y=j;x>=0 && y<n;x--,y++){
        if(a[x][y]==1){
            return false;
        }
    }
    return true;
    
}
bool solvenqueens(int a[][20],int n,int i){
    //base case
    if(i==n){
        return true;
    }
    
    //rec case
    for(int x=0;x<n;x++){
        if(canplace(a,n,i,x)){
            a[i][x]=1;
            if(solvenqueens(a,n,i+1)){
                return true;
            }
            a[i][x]=0;
            
        }
    }
    return false;
}


int main()
{
    int n;
    cin>>n;
    int a[20][20]={0};
    if(solvenqueens(a,n,0)){
    printboard(a,n);
    }
    else{
        cout<<"no solution exists"<<endl;
    }
    return 0;
}

File: SDE/Python/Practice/NQueens.py
board = 0
n = 0


def print_board():
    for v in board:
        for i in v:
            if i == 0:
                print('*', end=" ")
            else:
                print('Q', end=" ")
        print('\n')
#00 01 02 03 04
#10 11 12 13 14
#20 21 22 23 24
#30 31 32 33 34

def canplace(i,j):
    #check up
    x = i
    y = j
    while  x>=0 :
        if board[x][y] ==1:
            return False
        x-=1

    #check left up
    x = i
    y = j
    while x >= 0 and y >= 0:
        if board[x][y] == 1:
            return False
        x -= 1
        y -= 1
    #chec right up
    x = i
    y = j
    while x >= 0 and y < n:
        if board[x][y] == 1:
            return False
        x -= 1
        y += 1
    return True

def solvenqueens(i):
    if i>=n:
        return True
    for k in range(0,n):
        if(canplace(i,k)):
            board[i][k]= 1
            if(solvenqueens(i+1)):
                return True
            board[i][k]=0
    return False

n = int(input(('enter number of Queens..')))
print()
board = [[0 for _ in range(n)] for _ in range(n)]


if(solvenqueens(0)):
    print_board()
else:
    print("solution not possible")
   # print_board()

